### DISCUSSION TITLE
Tower Defense - Class-Diagram

### DISCUSSION THREAD

**Discussion by argastle** [Freaky2164/TowerDefense#9] [General]
Welcome back to our blog,

This week we created a class-diagram as a first draw for the implementation. 

![Class-Diagram](https://user-images.githubusercontent.com/97842228/199693287-d421d9de-1e22-4318-9c1d-8de578091830.png)

It was not that easy to apply the principles presented to us, due to the nature of Unity. So the only principle from solid we were able to apply was the single responsibility principle.

We tried to isolate the functionalities so that every class is only responsible for the actions it is required to do. 
For example the tower is only responsible for attacking enemies, but does not have to know, which one he is going to attack, that is up to the projectile.

Disclaimer
Caused by our lack of experience with Unity, the diagram above is only a rough sketch and will most likely change heavily in the future.

For our next sprint we will try to start with some basic implementation.

~ The TowerDefence Team

**Comment by statzie** [comment]
Hey 🏰🛡️,

overall your diagram looks very well structured and is easy to read.
Our only question is about your compositions that reference the same class they originate from.
We are not sure if it is possible, as we have never seen it before.
What was meaning of these compositions?

Much love from the RALFers,
^oe

**Reply by TimWaeckerle** [reply]
Hi,

since the object is able to create an instance of itself in an upgraded version we choose them to be in a relationship.

Thanks for you r contribution!

TD Team

**Comment by L2S8** [comment]
Hello TowerDefense-Team,

your UML-diagram is well made and considers all the important classes and relationships. The core logic of the game mechanics are well thought through already. 
Have you been planning on adding different types of towers, or projectiles? In that case maybe a "type" attribute would fit into these classes aswell.
On a personal note, I've always enjoyed TowerDefense games, so I'm quite interested in the outcome of your project and what unique twist you might give it.

-Common-Team

**Reply by TimWaeckerle** [reply]
Hi,

we don t have a type attribute on a turret because we got diffrent gameobjects of turrets which contain the used projectile.
This is a common use of Unity objects.

TD Team
